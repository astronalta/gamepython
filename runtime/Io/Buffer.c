/*
 * Copyright (c) 2010 Matt Fichman
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, APEXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

#include "Io/Buffer.h"
#include <stdlib.h>
#include <stdio.h>


Io_Buffer Io_Buffer__init(Int capacity) {
    Io_Buffer ret = calloc(sizeof(struct Io_Buffer) + capacity, 1);
    if (!ret) {
        fprintf(stderr, "Out of memory");
        fflush(stderr);
        abort();
    }
    ret->_vtable = Io_Buffer__vtable;
    ret->_refcount = 1;
    ret->capacity = capacity;
    ret->begin = 0;
    ret->end = 0;
    return ret; 
}

Byte Io_Buffer__index(Io_Buffer self, Int index) {
    if (index < self->begin || index >= self->end) {
        return 0;
    } else {
        return self->data[self->begin+index];
    }
}

void Io_Buffer__insert(Io_Buffer self, Int index, Byte byte) {
    if (index < self->begin || index >= self->end) {
        return;
    } else {
        self->data[index] = byte;
    }
}

