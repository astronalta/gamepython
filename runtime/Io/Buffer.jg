# Copyright (c) 2010 Matt Fichman
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, APEXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

Buffer < Object {
    # Container for a static array of bytes.  This class is implemented using 
    # the 'C struct hack', so allocating a buffer of a certain size will 
    # allocate a contiguous block of memory for this object.  As such, the
    # actual runtime size of this object depends on the capacity value passed
    # to the constructor.
    
    capacity Int
    # Returns the capacity of this buffer.

    begin mutable Int
    # Returns the index of first valid byte in this buffer.

    end mutable Int
    # Returns the index after the last valid byte in this buffer.

    @init(size Int) native {
        # Creates a new buffer with capacity for 'size' bytes.
    }

    @insert(index Int, byte Byte) native {
        # Sets the byte at 'index' to 'byte'.
    }

    @index(index Int) native Byte {
        # Returns the byte at 'index'. 
    }

    begin=(value Int) {
        # Sets the index of the first byte stored in this buffer.  Note that
        # this does not change the capacity.  Setting this value to a number
        # greater than the capacity or 'end' results in 'begin' being set to
        # the minimum of those two values.
        begin = value.min(capacity).min(end).max(0)
    }

    end=(value Int) {
        # Sets the index following the last byte stored in this buffer.
        end = value.min(capacity+1).max(begin).max(0)
    }

    empty?() Bool {
        # Returns true if the buffer is completely empty.
        ret begin == end
    }

    full?() Bool {
        # Returns true if there is no more space to write in this buffer.
        ret end == capacity 
    }
    
}
