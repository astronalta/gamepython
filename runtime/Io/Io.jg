# Copyright (c) 2010 Matt Fichman
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, APEXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

println(str String) {
    # Prints 'str' to standard output, and flushes the output.
    if str {
        stdout().println(str)
    } else {
        stdout().println('nil')
    }
}

print(str String) {
    # Prints 'str' to standard output.  Does not flush the output.
    if str {
        stdout().print(str)
    } else {
        stdout().print('nil')
    }
}

scanln() String {
    # Scans all input up to the end of the line.  Reads the end-of-line
    # character, but doesn't return it.
    ret stdin().scanln()
}

scan(str String) String {
    # Scans for the delimiter specified by 'str' and returns all input up to
    # the delimiter.
    ret stdin().scan(str)
}

stderr() native Stream {
    # Returns the standard error stream
}

stdout() native Stream {
    # Returns the standard output stream
}

stdin() native Stream {
    # Returns the standard input stream
}

manager() native Manager {
    # Returns the default Io::Manager
}
