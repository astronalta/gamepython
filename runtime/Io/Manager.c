
/*
 * Copyright (c) 2010 Matt Fichman
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, APEXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

#include "Io/Manager.h"
#include "Coroutine.h"
#ifndef WINDOWS
#include <unistd.h>
#else
#include <windows.h>
#endif
#include <stdlib.h>
#include <stdio.h>


Io_Manager Io_Manager__init() {
    // Initialize an event manager, and allocate an I/O completion port.
    Io_Manager ret = calloc(sizeof(struct Io_Manager), 1);
    if (!ret) {
        fprintf(stderr, "Out of memory");
        fflush(stderr);
        abort();
    }
    ret->_vtable = Io_Manager__vtable;
    ret->_refcount = 1;
    ret->scheduled = Queue__init(0);
#ifdef WINDOWS
    ret->handle = (Int)CreateIoCompletionPort(INVALID_HANDLE_VALUE, 0, 0, 1);
#else
    ret->handle = 0;
#endif
    return ret;
}

void Io_Manager__destroy(Io_Manager self) {
    Queue__destroy(self->scheduled);
#ifdef WINDOWS
    CloseHandle((HANDLE)self->handle);
#else
#endif
    free(self);
}

void Io_Manager_poll(Io_Manager self) {
    // Poll for an I/O event, and then resume the coroutine associated with
    // that event.
#ifdef WINDOWS
    DWORD bytes = 0;
    ULONG_PTR coro = 0;
    OVERLAPPED* evt = 0;
    HANDLE handle = (HANDLE)self->handle;
    GetQueuedCompletionStatus(handle, &bytes, &coro, &evt, INFINITE);
    if (evt) {
        Coroutine__call((Coroutine)coro); // Will return later
    }
#endif
}

