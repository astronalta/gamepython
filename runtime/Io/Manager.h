/*
 * Copyright (c) 2010 Matt Fichman
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, APEXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

#ifndef JOGO_IO_MANAGER_H
#define JOGO_IO_MANAGER_H

#include "Primitives.h"
#include "Coroutine.h"
#include "Queue.h"
#ifdef WINDOWS
#include <windows.h>
#endif

typedef struct Io_Manager* Io_Manager;
struct Io_Manager {
    Ptr _vtable;
    U64 _refcount;
    Queue scheduled;
    Int handle;
    Int waiting;
    Int iobytes;
};

#ifdef WINDOWS
typedef struct Io_Overlapped {
    OVERLAPPED overlapped;
    Coroutine coroutine;
} Io_Overlapped;
#endif

Io_Manager Io_Manager__init();
void Io_Manager__destroy();
void Io_Manager_poll(Io_Manager self);
void Io_Manager_shutdown(Io_Manager self);
Int Io_manager_iobytes__g(Io_Manager self);
extern void Io_Manager__vtable();

void Io_wait();
Io_Manager Io_manager();

#endif
