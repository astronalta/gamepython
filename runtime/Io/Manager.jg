# Copyright (c) 2010 Matt Fichman
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, APEXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

Manager < Object {
    # The Io::Manager coordinates coroutines that have blocked waiting for I/O,
    # and provides the core asynchronous I/O implementation.  The Io::Manager is
    # resposible for running an Jogo program's event loop; the Io::Manager is 
    # started first by the runtime, and the the main() coroutine is launched
    # and added to the scheduled task queue.  
    # 
    # Arbitrary tasks (named coroutines) can be added to the task queue.  When
    # a task is added to the task queue, it may be run the next time the
    # current coroutine yields or blocks on I/O.   When a task has finished
    # running, the task is removed from the task queue. 

    scheduled Queue[Coroutine]
    # Set of tasks that are active and waiting to be scheduled.

    handle Int
    # Handle to the file descriptor that is used for OS-level polling.  On
    # BSD-based systems, it is a kqueue fd; for Linux, it is an epoll fd; and
    # for Windows, it is an I/O completion port.

    waiting private Int
    # Number of coroutines waiting on I/O.  FixMe: Make this a set of waiting
    # coroutines rather than a single int.

    iobytes immutable Int
    # Number of bytes read by the last I/O operation, if available.

    @init() native private {
        # Initializes a new manager.  Most applications need only one I/O
        # manager, which can be accessed via the Event::manager() constant.
    }
    
    @destroy() native {
        # Destroys the I/O manager, and releases the OS resources used for
        # event polling.
    }

    task(task Coroutine) {
        # Adds this task to the list of tasks waiting to be scheduled.  The
        # task will be run when all other tasks are blocked or if a task yields
        # and the task is first on the task list.  If the task blocks to do
        # I/O, it will yield back to the I/O manager, which will schedule
        # another task.  If no tasks are available, then the I/O manager will
        # wait on I/O.
        scheduled.enq(task)
    }

    run() {
        # Run the event loop.  Select a coroutine from the 'scheduled' queue,
        # and begin executing it.  If there are no more tasks, and no
        # outstanding I/O requests, then return.
        while waiting > 0 {
            # Dequeue scheduled tasks until the entire queue is empty.  At that
            # point, all coroutines have either exited, or are waiting on some
            # external event (file I/O, timer, etc.)
            while scheduled.count > 0 {
                next = scheduled.deq()
                next.resume()
                if next.status == CoroutineStatus::SUSPENDED {
                    scheduled.enq(next)
                }
            }
        
            # All coroutines are sleeping; poll.
            poll()
        }
    }

    poll() native {
        # Poll for an I/O request.  This function will block until an I/O
        # operation finishes.
    }

    shutdown() native {
        # Shuts down the I/O manager; no additional I/O operations can be done
        # after this method is called.
    }
}

