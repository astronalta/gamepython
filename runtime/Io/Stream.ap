# Copyright (c) 2010 Matt Fichman
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
# 
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, APEXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

StreamStatus = OK | ERROR | EOF
# Encodes the current status of the I/O stream.  OK: Stream is in the normal 
# state and is ready for I/O.  ERROR: The stream is in an invalid state, either
# because a socket or file descriptor is closed, or due to an error raised by
# the underlying file descriptor.  EOF: The stream is at the end of the file,
# and no more bytes may be read.

StreamMode = ASYNC | BLOCKING
# The stream mode manages how I/O is handled by the runtime.  ASYNC: Stream
# read/write operations return immediately, but the I/O operation will not
# finish until later; the wait() function must be called to retrieve the stream
# status.  SYNC: Stream operations block the current coroutine.  BLOCKING:
# Stream operations block the process.  SYNC is the default.  BLOCKING mode is
# not recommended for most applications.

StreamType = FILE | SOCKET | CONSOLE
# The type of underlying file descriptor that is attached to the stream.

Stream < Object {
    # This object is used to perform buffered character I/O on a file.  It can
    # be used to read from any device or input that the OS supports for the
    # underlying handle type, including sockets, files, pipes, etc.

    handle immutable Int
    # Opaque handle to the underlying file descriptor.

    read_buf private Buffer
    # Bytes buffered for input.  The stream buffers input bytes to avoid 
    # invoking a system call for each read.

    write_buf private Buffer
    # Bytes buffered for output.  The stream buffers output bytes to avoid 
    # invoking a system call for each write.

    status StreamStatus
    # The current status of the stream; either OK, ERROR, or EOF.

    mode StreamMode
    # The current mode of the stream, as described above; either SYNC, ASYNC,
    # or BLOCKING.  Controls the type of I/O that will be executed for the 
    # stream.

    type immutable StreamType
    # True if the stream is a console handle    

    coroutine private Coroutine
    # Coroutine that is blocked waiting for I/O from this stream

    @init(desc Int, type StreamType) native {
        # Initializes a new stream, with 'handle' being an opaque pointer to
        # the OS file descriptor.  Note that if 'handle' is invalid, then
        # subsequent reads and writes will fail, and 'err' will be set.
    }

    @destroy() {
        # Closes the file if it is still open.
        self.close()
    }

    read(buffer Buffer) native {
        # Reads data into 'buffer', and sets the 'count' field of buffer to the
        # number of bytes read.  This function will read up 'n' bytes, where
        # 'n' is equal to the 'length' field of 'buffer'.  If data is already
        # buffered to the internal buffer, that data will be copied into
        # 'buffer' first before an attempt to read more bytes from the OS is
        # made.  This function implements an unbuffered read; the bytes
        # returned by the OS-level read call go directly into the buffer.
    }

    write(buffer Buffer) native {
        # Writes the data in 'buffer', and blocks until all the bytes are
        # written or the 'err' flag is set.  If data is already buffered in the
        # internal buffer, then 'flush()' will be called before writing
        # 'buffer' to the output stream.  This function implements an
        # unbuffered write; the bytes will go directly to the OS-level write
        # system call.
    }

    get() native Char {
        # Returns the next character in the input stream. 
    }

    put(char Char) native {
        # Writes 'char' to the output buffer.  The character is not guaranteed
        # to be written to the output stream until 'flush()' is called.
    }
    
    scan(delim String) native String {
        # Reads all characters up to and including the first instance of
        # 'char'.  Returns a string containing the characters read.
    }

    print(string String) native {
        # Writes the characters in 'string' to the file.  The string is not
        # guaranteed to be written to the output stream until 'flush()' is
        # called.
    }

    scanln() String {
        # Reads in a line up to the ending '\n'
        ret self.scan('\n')
    }

    println(string String) {
        # Prints a string followed by '\n', and flushes the output.
        self.print(string)
        self.put('\n'c)
        self.flush()
    }
    
    peek() native Char {
        # Reads a character, but does not remove it from the internal buffer.
        # Call 'get' to read the next character and remove it.
    }

    flush() native {
        # Writes any characters in the internal output buffer to the output
        # stream.
    }
    
    close() native {
        # Calls 'flush()' then closes the file descriptor, so that no more
        # characters may be read or written.
    }
}
