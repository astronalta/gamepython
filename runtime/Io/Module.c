/*
 * Copyright (c) 2010 Matt Fichman
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, APEXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

#include "Io/Stream.h"
#include "String.h"
#include <stdio.h>

void Io_println(String str) {
    // FIXME: Use Apollo streams
    if (str) {
        fwrite(str->data, str->length, 1, stdout);
    } else {
        fwrite("nil", 3, 1, stdout);
    }
    fwrite("\n", 1, 1, stdout);
    fflush(stdout);
}

void Io_print(String str) {
    // FIXME: Use Apollo streams
    if (str) {
        fwrite(str->data, str->length, 1, stdout);
    } else {
        fwrite("nil", 3, 1, stdout);
    }
    fflush(stdout);
}
