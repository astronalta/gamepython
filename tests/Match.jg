# jgtest exec on

import Io

Fruit = APPLE | ORANGE | PEAR | BANANA | STRAWBERRY | KIWI

switch(fruit Fruit) {
    match fruit {
    with Fruit::APPLE: print('apple') 
    with Fruit::ORANGE: print('orange')
    with Fruit::PEAR: print('pear')
    with Fruit::BANANA: print('banana') 
    with Fruit::STRAWBERRY: print('strawberry')
    with *: print('unknown')
    }
    println('.')
}

switch2(fruit Fruit) {
    match fruit {
    with Fruit::APPLE: 
        print('apple') 
        print('1')
    with Fruit::ORANGE: 
        print('orange')
        print('2')
    with Fruit::PEAR: 
        print('pear')
        print('3')
    with Fruit::BANANA: 
        print('banana') 
        print('4')
    with *: 
        print('unknown')
        print('5')
    }
    println('.')
}

main() Int {

    switch(Fruit::APPLE)
    switch(Fruit::ORANGE)
    switch(Fruit::PEAR)
    switch(Fruit::BANANA)
    switch(Fruit::STRAWBERRY)
    switch(Fruit::KIWI)
    
    switch2(Fruit::APPLE)
    switch2(Fruit::ORANGE)
    switch2(Fruit::PEAR)
    switch2(Fruit::BANANA)
    switch2(Fruit::STRAWBERRY)

    ret 0
}
