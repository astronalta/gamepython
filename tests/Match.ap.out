apple.
orange.
pear.
banana.
strawberry.
apple1.
orange2.
pear3.
banana4.
.
